dev_scripts/env.py

```diff
@@ -0,0 +1,503 @@
+#!/usr/bin/env python3
+
+import argparse
+import getpass
+import os
+import pathlib
+import shutil
+import subprocess
+import sys
+
+DEFAULT_GUI = False
+DEFAULT_USER = "user"
+DEFAULT_DRY = False
+DEFAULT_DEV = False
+DEFAULT_SHOW_DOCKERFILE = False
+
+# The Linux distributions that we currently support.
+# FIXME: Add a version mapping to avoid mistakes.
+# FIXME: Maybe create an enum for these values.
+DISTROS = ["debian", "fedora", "ubuntu"]
+CONTAINER_RUNTIMES = ["podman", "docker"]
+IMAGE_NAME_BUILD_DEV_FMT = "dangerzone.rocks/build/{distro}:{version}"
+IMAGE_NAME_BUILD_FMT = "dangerzone.rocks/{distro}:{version}"
+
+EPILOG = """\
+Examples:
+
+Build a Dangerzone environment for development (build-dev) or testing (build) based on
+Ubuntu 22.04:
+
+    env.py --distro ubuntu --version 22.04 build-dev
+    env.py --distro ubuntu --version 22.04 build
+
+Inspect the Dockerfile for the environments:
+
+    env.py --distro ubuntu --version 22.04 build-dev --show-dockerfile
+    env.py --distro ubuntu --version 22.04 build --show-dockerfile
+
+Run an interactive shell in the development or end-user environment:
+
+    env.py --distro ubuntu --version 22.04 run --dev bash
+    env.py --distro ubuntu --version 22.04 run bash
+
+Run Dangerzone in the development environment:
+
+    env.py --distro ubuntu --version 22.04 run --dev --gui bash
+    user@dangerzone-dev:~$ cd dangerzone/
+    user@dangerzone-dev:~$ poetry run ./dev/scripts/dangerzone
+
+Run Dangerzone in the end-user environment:
+
+    env.py --distro ubuntu --version 22.04 run --gui dangerzone
+
+"""
+
+DOCKERFILE_BUILD_DEV_DEBIAN_DEPS = rf"""
+RUN apt-get update && apt-get install -y \
+    podman dh-python python3 python3-stdeb python3-pyside2.qtcore \
+    python3-pyside2.qtgui python3-pyside2.qtwidgets python3-appdirs \
+    python3-click python3-xdg python3-colorama
+
+RUN apt-get update && apt-get install -y make
+RUN apt-get update && \
+    DEBIAN_FRONTEND=noninteractive apt-get install -y \
+    python3-dev python3-venv python3-pip
```

Comment:
Hm, I'm not sold on the above suggestions because I think (1) breaks caching, which we get for free with `docker/podman build` and (2) would make `env.py` more complex.

Yet I totally feel the pain. We need to do something about it, but I hoped we could punt it for the future. The reason is that the problem seems to me much bigger than the drift between `env.py` / `BUILD.md`. We already have lots of places where we define dependencies. Here's a (hopefully) exhaustive list:

* `.circleci/config.yml`
* `BUILD.md`
* `dev_scripts/env.py`
* `install/linux/build-rpm.py`
* `pyproject.toml`
* `stdeb.cfg`

I think we need to approach this a bit differently:

1. What is the role of these dependencies? How do they differ from the ones already specified in `pyproject.toml`?

    My guess is that they are the system equivalent dependencies of the Python ones, and they are there prior to Poetry's introduction (see this [past README.md version](https://github.com/freedomofpress/dangerzone/blob/8a84c115a1af78bcee62bfc8d255a7ff3e99efc7/README.md)).

2. Can we ditch them and just ask the user to do `poetry install ...`?

     I think yes. The main problem is that the development environment will not be on par with the user's environment, but that's not a problem. Part of our QA steps is to check the **installed* Dangerzone application in a clean environment (which `env.py` already provides), so we will catch any errors there.

3. What about `stdeb.cfg` and `install/linux/build-rpm.py`, which refer to the system dependencies?

     Both these files are used by Python code. We can extend this Python code to check that the source of truth (`pyproject.toml`) has no diverged since writing the list of system dependencies. Similar to what you did with the reference trick.

What do you think?

Reply:
After a synchronous discussion, these were our conclusions:

1. Remove deps from .circleci/config.yml / BUILD.md / env.py
2. Move any deps that are not listed on pyproject toml (stdeb?) to pyproject.toml, preferably under a different section (e.g., tools.poetry.group.package).
3. Test these changes.
4. Create an issue for removing the system deps from stdeb.cfg and build-rpm.py (e.g, dh-python could read them from pyproject.toml) (optional)